waiter/waiter.go

```diff
@@ -0,0 +1,77 @@
+package waiter
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"time"
+)
+
+const (
+	// DefaultPollInterval represents duration between checks. Each DefaultPollInterval waiter make the request
+	// to confirm the operation it waits, was successfully executed.
+	// For instance: ContainerPutWaiter waits until container will be created.
+	DefaultPollInterval = 1 * time.Second
+)
+
+var (
+	// ErrConfirmationTimeout describes situation, when we sent request, but got timeout during waiting confirmation.
+	// For instance: we send containerPut, after that we wait and do containerGet, to be sure it was created successfully.
+	// If we have context expired at this step, we return this error.
+	ErrConfirmationTimeout = errors.New("confirmation timeout")
+)
+
+// Executor describes requirements for async.
+type Executor interface {
+	ContainerDeleteExecutor
+	ContainerSetEACLExecutor
+	ContainerPutExecutor
+}
+
+// Waiter combines async [client.Client]/[pool.Pool] methods and gives sync alternative of them with the same func signatures.
+type Waiter struct {
+	ContainerPutWaiter
+	ContainerSetEACLWaiter
+	ContainerDeleteWaiter
+}
+
+// The function implements poll logic for each waiter.
+//
+// The first return value is a flag:
+//   - If true, the operation executed successfully, stop waiter.
+//   - If false, the operation still in progress, wait one more turn.
+//
+// The second return value is an error: If err != nil, stop waiter immediately and return this error.
+type pollLogic func() (bool, error)
+
+// NewWaiter is a constructor for [waiter.Waiter].
+func NewWaiter(executor Executor, pollInterval time.Duration) *Waiter {
+	w := &Waiter{
+		ContainerPutWaiter:     NewContainerPutWaiter(executor, pollInterval),
+		ContainerSetEACLWaiter: NewContainerSetEACLWaiter(executor, pollInterval),
+		ContainerDeleteWaiter:  NewContainerDeleteWaiter(executor, pollInterval),
+	}
+
+	return w
+}
+
+func poll(ctx context.Context, pollInterval time.Duration, callBack pollLogic) error {
+	t := time.NewTicker(pollInterval)
+	defer t.Stop()
+
+	for {
+		select {
+		case <-t.C:
+			isDone, err := callBack()
+			if err != nil {
+				return fmt.Errorf("poll: %w", err)
+			}
+
+			if isDone {
+				return nil
```

Comment:
Because we can be done, but have an error (a fatal one). Or can be done, but not have it (success!). Otherwise a single `err` returned from the callback would be enough.

Reply:
Which can be achieved btw if `callback` is to return some specific `errRetry`. Maybe this would even be more readable this way (single error return, retries are easily seen in the code and anything else is either a fatal problem or a total success).

Reply:
`errRetry` is more clear and readable, refactored

Reply:
Yeah, really much better this way.